File: leetcode/Arrays/leetcode-0015/15.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        if (nums == null || nums.length < 3) {
            return res;
        }

        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            int low = i + 1, high = nums.length - 1;
            while (low < high) {
                if (nums[low] + nums[high] + nums[i] == 0) {
                    res.add(Arrays.asList(nums[i], nums[low], nums[high]));
                    while (low + 1 < high && nums[low] == nums[low + 1]) {
                        low++;
                    }
                    while (low < high - 1 && nums[high] == nums[high - 1]) {
                        high--;
                    }
                    low++;
                    high--;
                } else if (nums[low] + nums[high] + nums[i] < 0) {
                    low++;
                } else {
                    high--;
                }
            }
        }
        return res;
    }
}

File: leetcode/Arrays/leetcode-0015/15.go
package arrays

import "sort"

func threeSum(nums []int) [][]int {
	res := [][]int{}
	if len(nums) < 3 {
		return res
	}

	sort.Ints(nums)

	for i := 0; i < len(nums)-2; i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		low, high := i+1, len(nums)-1
		for low < high {
			if nums[i]+nums[low]+nums[high] == 0 {
				res = append(res, []int{nums[i], nums[low], nums[high]})
				for low < high && nums[low] == nums[low+1] {
					low++
				}
				for low < high && nums[high] == nums[high-1] {
					high--
				}
				low++
				high--
			} else if nums[i]+nums[low]+nums[high] < 0 {
				low++
			} else {
				high--
			}
		}
	}
	return res
}
